From fe78940e6fb5f0ea1d9e776f80899816611167fa Mon Sep 17 00:00:00 2001
From: "David A. Wheeler" <dwheeler@dwheeler.com>
Date: Sun, 22 Sep 2019 15:22:17 -0400
Subject: [PATCH] Add better error message on encoding problems

Python3 doesn't provide easy-to-use built-in libraries to deal
with common encoding issues (e.g., Windows-1252 encoded characters
in a UTF-8 stream), so when we see an encoding error,
provide better information on how to deal with it and
a pointer to the more detailed information in the documentation.

Signed-off-by: David A. Wheeler <dwheeler@dwheeler.com>
---
 flawfinder | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/flawfinder b/flawfinder
index 553fa31..bbccdbc 100755
--- a/flawfinder
+++ b/flawfinder
@@ -1505,7 +1505,19 @@ def process_c_file(f, patch_infos):
             print("Examining", f)
         sys.stdout.flush()
 
-    text = "".join(my_input.readlines())
+    # Python3 is often configured to use only UTF-8, and presumes
+    # that inputs cannot have encoding errors.
+    # The real world isn't like that, so provide a prettier warning
+    # in such cases - with some hints on how to solve it.
+    try:
+        text = "".join(my_input.readlines())
+    except UnicodeDecodeError as err:
+        print('Error: encoding error in', h(f))
+        print(err)
+        print('Run as PYTHONUTF8=0 LC_ALL=C.ISO-8859-1 python3 flawfinder,')
+        print('convert source code to UTF-8, or run flawfinder using python2.')
+        print('See documentation for more information.')
+        sys.exit(15)
 
     i = 0
     while i < len(text):
